src/databricks/labs/ucx/assessment/azure.py

```diff
@@ -23,161 +23,113 @@
-            matched = matching_key[0]
-            if not matched:
-                return None
-            if re.search("spark_conf", matched):
-                client_endpoint_list = config.get(matched, {}).get("value", "").split("/")
-            else:
-                client_endpoint_list = config.get(matched, "").split("/")
-            if len(client_endpoint_list) == _CLIENT_ENDPOINT_LENGTH:
-                return client_endpoint_list[3]
-        return None
-
-    def _get_azure_spn_list(self, config: dict) -> list:
-        spn_list = []
-        secret_scope, secret_key, tenant_id, storage_account = None, None, None, None
-        matching_key_list = [key for key in config.keys() if "fs.azure.account.oauth2.client.id" in key]
-        if len(matching_key_list) > 0:
-            for key in matching_key_list:
-                storage_account_match = re.search(_STORAGE_ACCOUNT_EXTRACT_PATTERN, key)
-                if re.search("spark_conf", key):
-                    spn_application_id = config.get(key, {}).get("value")
-                else:
-                    spn_application_id = config.get(key)
-                if not spn_application_id:
-                    continue
-                secret_matched = re.search(_SECRET_PATTERN, spn_application_id)
-                if secret_matched:
-                    spn_application_id = self._check_secret_and_get_application_id(secret_matched)
-                    if not spn_application_id:
-                        continue
-                    secret_scope, secret_key = (
-                        secret_matched.group(1).split("/")[1],
-                        secret_matched.group(1).split("/")[2],
-                    )
-                if storage_account_match:
-                    storage_account = storage_account_match.group(1).strip(".")
-                    tenant_key = "fs.azure.account.oauth2.client.endpoint." + storage_account
-                else:
-                    tenant_key = "fs.azure.account.oauth2.client.endpoint"
-                tenant_id = self._get_azure_spn_tenant_id(config, tenant_key)
-
-                spn_application_id = "" if spn_application_id is None else spn_application_id
-                secret_scope = "" if secret_scope is None else secret_scope
-                secret_key = "" if secret_key is None else secret_key
-                tenant_id = "" if tenant_id is None else tenant_id
-                storage_account = "" if storage_account is None else storage_account
-                spn_list.append(
-                    {
-                        "application_id": spn_application_id,
-                        "secret_scope": secret_scope,
-                        "secret_key": secret_key,
-                        "tenant_id": tenant_id,
-                        "storage_account": storage_account,
-                    }
-                )
-        return spn_list
+        return list(generate_service_principals(deduped_service_principals))
```

Comment:
if you make a set of `AzureServicePrincipalInfo` and return those directly instead of working with dictionaries - you don't need a dedupe block - dataclasses already have `__hash__`.